src/SDK/Trace/TracerProviderFactory.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\SDK\Trace;
+
+use OpenTelemetry\API\Trace as API;
+
+final class TracerProviderFactory
+{
+    private string $name;
+    private ExporterFactory $exporterFactory;
+    private SamplerFactory $samplerFactory;
+    private SpanProcessorFactory $spanProcessorFactory;
+
+    public function __construct(
+        string $name,
+        ?ExporterFactory $exporterFactory = null,
+        ?SamplerFactory $samplerFactory = null,
+        ?SpanProcessorFactory $spanProcessorFactory = null
+    ) {
+        $this->name = $name;
+        $this->exporterFactory = $exporterFactory ?: new ExporterFactory($this->name);
+        $this->samplerFactory = $samplerFactory ?: new SamplerFactory();
+        $this->spanProcessorFactory = $spanProcessorFactory ?: new SpanProcessorFactory();
+    }
+
+    public function create(): API\TracerProviderInterface
```

Comment:
The semantics of the method suggest, that this is the only way to create a TracerProvider, while there could be possibly more. So either this class could be renamed appropriatly or the method. 
Type hinting `TracerProviderInterface` suggests, that there can be possibly more factories, and this factory is implementing an interface.

Reply:
My approach to type hinting is that it's documentation and safety (and it also helps IDEs), and if a method's return value is known then it should be type-hinted. I'm not sure why I wouldn't type-hint it, in theory we could create a second TracerProvider in the future and this factory would take care of determining which one to create...?

Reply:
I was not wondering _why_ you use a type hint, but why it's TracerProviderInterface and not TracerProvider in this case. And since it's TracerProviderInterface, there could be a TracerProviderFactoryInterface as well. So this factory could implement the interface and we could use other factories using other methods in the future. (Not saying, that this has to be done in this PR though.)

Reply:
I see - I guess because I think it's better to code to an interface rather than a concrete implementation, and this factory could in theory create other trace providers, based on inputs that we have not invented yet!

Reply:
Ok, got you.
Just for the record. I had also https://github.com/open-telemetry/opentelemetry-php/issues/301 in mind, while this is aiming at even another level above (SDK factory).